### PR TITLE
Issue 138

### PR DESCRIPTION
**Pull request by Zqy11**
Complete issue #138 and fix all issues for new compile warnings.

### CODE REVIEW

File: include/aca_util.h [hunk 1]
```diff
@@ -116,7 +116,7 @@ static inline bool aca_validate_mac_address(const char *mac_string)
   unsigned char mac[6];
 
   if (mac_string == nullptr) {
-    ACA_LOG_ERROR("Input mac_string is null\n");
+    ACA_LOG_ERROR("%s", "Input mac_string is null\n");
```
Comment (er1cthe0ne) [inline include/aca_util.h:119]: what is the compiler warning for this? the change looks a bit weird...

Note that we plan to update the logging system in ACA tracked by #116

File: src/ovs/aca_ovs_control.cpp [hunk 2]
```diff
@@ -56,11 +56,13 @@ int ACA_OVS_Control::control()
 {
   int overall_rc = EXIT_SUCCESS;
 
-  char target[g_ofctl_target.size() + 1];
+  // char target[g_ofctl_target.size() + 1];
+  char *target = new char[g_ofctl_target.size() + 1];
```
Comment (er1cthe0ne) [inline src/ovs/aca_ovs_control.cpp:60]: interesting to see compiler flagged this as warning. keep in mind you will need to free the newly allocated memory before function exit.

@cj-chung - please review this to make sure the change is correct

File: src/ovs/aca_ovs_control.cpp [hunk 3]
```diff
@@ -56,11 +56,13 @@ int ACA_OVS_Control::control()
...
   g_ofctl_target.copy(target, g_ofctl_target.size() + 1);
   target[g_ofctl_target.size()] = '\0';
 
-  char options[g_ofctl_options.size() + 1];
+  // char options[g_ofctl_options.size() + 1];
+  char *options = new char[g_ofctl_options.size() + 1];
```
Comment (er1cthe0ne) [inline src/ovs/aca_ovs_control.cpp:65]: same as above

**Comment by Zqy11** [comment]
The warning is like the first figure.




Because the variadic macros defined in the .../include/aca_log.h has two arguments.



And I find a post in stackoverflow.
https://stackoverflow.com/questions/43991106/iso-c99-requires-at-least-one-argument-for-the-in-a-variadic-macro-when







------------------ Original ------------------

**Review by er1cthe0ne — COMMENTED** [review]
looking good. please revert the permission changes on the below files because it is causing CI test to break (CI tests needs the current file permission to execute the scripts):

build/aca-machine-init.sh 
build/build.sh

**Comment by Zqy11** [comment]
OK ！


As I am in the dormitory now and have no computer on hand, I will submit it again when I arrive at the laboratory tomorrow.


------------------ Original ------------------

File: src/ovs/aca_ovs_control.cpp (hunk 2 above)
Comment (cj-chung) [reply src/ovs/aca_ovs_control.cpp:60]: If using `new` to create a `char*`, should we explicitly to `delete[] target`?

**Review by er1cthe0ne — COMMENTED** [review]
we are almost there, with two remaining items:

1. you will need to free/delete the newly allocated memory for char* in your two changes.
2. we don't need to change the permission for build/Dockerfile and build/Dockerfile.proxy, please revert your change on these two files.

File: include/aca_dhcp_server.h [hunk 4]
```diff
@@ -123,7 +123,8 @@ struct dhcp_client_id {
   uint8_t code;
   uint8_t len;
   uint8_t type;
-  uint8_t cid[0];
+  // uint8_t cid[0];
+  uint8_t *cid = new uint8_t[0];
```
Comment (er1cthe0ne) [inline include/aca_dhcp_server.h:127]: I just noticed this (sorry I didn't catch it previously), I am worry that we are allocating memory for a structure creation without the corresponding delete. 

What is the compiler warning for this? Can we surpress this warning if it doesn't make sense?

**Review by er1cthe0ne — COMMENTED** [review]
I just left a last comment. Please take a look. The last thing to do is to run the aca functional/unit tests to make sure all the 30 tests are passing with your change: ./build/tests/aca_tests

File: include/aca_dhcp_server.h (hunk 4 above)
Comment (Zqy11) [reply include/aca_dhcp_server.h:127]: The warning is like this:

In file included from /root/alcor-control-agent/test/gtest/aca_tests.cpp:28:0:
/root/alcor-control-agent/test/../include/aca_dhcp_server.h:126:16: warning: ISO C++ forbids zero-size array ‘cid’ [-Wpedantic]
   uint8_t cid[0];

-----------------------------------------------------------------------
The warning reason is that ISO C++ forbids zero-size array. I think we do not need to fix the warning (considering it wil bring with the problem of memory release).

**Comment by Zqy11** [comment]
I have run the test script "./build/tests/aca_tests" , and the output is the following:

[----------] Global test environment tear-down
[==========] 30 tests from 5 test suites ran. (2780 ms total)
[  PASSED  ] 30 tests.

I think all the 30 tests are passing with my changes.

File: include/aca_dhcp_server.h (hunk 4 above)
Comment (er1cthe0ne) [reply include/aca_dhcp_server.h:127]: agreed, let's surpress this warning then, please see the below link on the suggestions:

https://stackoverflow.com/questions/18843534/how-to-suppress-compiler-warning-c4373
https://www.fluentcpp.com/2019/08/30/how-to-disable-a-warning-in-cpp/

Comment (Zqy11) [reply include/aca_dhcp_server.h:127]: I can't find the warning code (such 4373) for gcc.
In fact, VC++'s warning code of "zero-size array" is 4200, but it doesn't work for gcc/make.

And I found that gcc has its own format for surpressing warnings.
Such as:
---------------------------------------
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
uint8_t cid[0];
#pragma GCC diagnostic pop
---------------------------------------

Comment (Zqy11) [reply include/aca_dhcp_server.h:127]: I can't find the warning code (such as 4373) for gcc.
In fact, VC++'s warning code of "zero-size array" is 4200, but it doesn't work for gcc/make.

And I found that gcc has its own format for surpressing warnings.
Such as:
---------------------------------------
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
uint8_t cid[0];
#pragma GCC diagnostic pop
---------------------------------------

Comment (Zqy11) [reply include/aca_dhcp_server.h:127]: I can't find the warning code (such 4373) for gcc.
In fact, VC++'s warning code of "zero-size array" is 4200, but it doesn't work for gcc/make.

And I found that gcc has its own format for surpressing warnings.
Such as:
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
uint8_t cid[0];
#pragma GCC diagnostic pop

**Pull request closed without merging**